lib/ansible/modules/web_infrastructure/acme_challenge_cert_helper.py

```diff
@@ -0,0 +1,271 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+# (c) 2018 Felix Fontein <felix@fontein.de>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+
+DOCUMENTATION = '''
+---
+module: acme_challenge_cert_helper
+author: "Felix Fontein (@felixfontein)"
+version_added: "2.7"
+short_description: Prepare certificates required for ACME challenges such as C(tls-alpn-01)
+description:
+   - "Prepares certificates for ACME challenges such as C(tls-alpn-01)."
+   - "The raw data is provided by the M(acme_certificate) module, and needs to be
+      converted to a certificate to be used for challenge validation. This module
+      provides a simple way to generate the required certificates."
+requirements:
+   - "cryptography >= 1.3"
+options:
+  challenge:
+    description:
+      - "The challenge type "
+    required: yes
+    choices:
+    - tls-alpn-01
+  challenge_data:
+    description:
+      - "The C(challenge_data) entry provided by M(acme_certificate) for the challenge."
+    required: yes
+  private_key_src:
+    description:
+      - "Path to a file containing the private key file to use for this challenge
+         certificate."
+      - "Mutually exclusive with C(private_key_content)."
+  private_key_content:
+    description:
+      - "Content of the private key to use for this challenge certificate."
+      - "Mutually exclusive with C(private_key_src)."
+'''
+
+EXAMPLES = '''
+- name: Create challenges for a given CRT for sample.com
+  acme_certificate:
+    account_key_src: /etc/pki/cert/private/account.key
+    challenge: tls-alpn-01
+    csr: /etc/pki/cert/csr/sample.com.csr
+    dest: /etc/httpd/ssl/sample.com.crt
+  register: sample_com_challenge
+
+- name: Create certificates for challenges
+  acme_challenge_cert_helper:
+    challenge: tls-alpn-01
+    challenge_data: "{{ item.value['tls-alpn-01'] }}"
+    private_key_src: /etc/pki/cert/key/sample.com.key
+  with_items: "{{ sample_com_challenge.challenge_data }}"
+  register: sample_com_challenge_certs
+
+- name: Install challenge certificates
+  # We need to set up HTTPS such that for the domain,
+  # regular_certificate is delivered for regular connections,
+  # except if ALPN selects the "acme-tls/1"; then, the
+  # challenge_certificate must be delivered.
+  # This can for example be achieved with very new versions
+  # of NGINX; search for ssl_preread and
+  # ssl_preread_alpn_protocols for information on how to
+  # route by ALPN protocol.
+  ...:
```

Comment:
leftover?

Reply:
No, not really, it is more a mark for "insert whatever module you need to do this here" ;-) (might require more than one module) How this task would look like depends a lot on the webserver of choice and how it is configured. I want to try this out with NGINX at some point to write a how-to on how this can actually be done, but so far I didn't got around to do that :)